{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Chapter 1: Mechanisms of Action (MoA) Prediction using DNN model"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Background\n",
    "\n",
    "The Connectivity Map, a project within the Broad Institute of MIT and Harvard, together with the Laboratory for Innovation Science at Harvard (LISH), presents a challenge [at kaggle](https://www.kaggle.com/c/lish-moa/overview) with the goal of advancing drug development through improvements to MoA prediction algorithms.\n",
    "\n",
    "We will be predicting multiple targets of the Mechanism of Action (MoA) response(s) of different samples (sig_id), given various inputs such as gene expression data and cell viability data.\n",
    "\n",
    "Two notes:\n",
    "\n",
    "- the training data has an additional (optional) set of MoA labels that are not included in the test data and not used for scoring.\n",
    "- the re-run dataset has approximately 4x the number of examples seen in the Public test."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### What is the Mechanism of Action (MoA) of a drug? And why is it important?\n",
    "\n",
    "In the past, scientists derived drugs from natural products or were inspired by traditional remedies. Very common drugs, such as paracetamol, known in the US as acetaminophen, were put into clinical use decades before the biological mechanisms driving their pharmacological activities were understood. Today, with the advent of more powerful technologies, drug discovery has changed from the serendipitous approaches of the past to a more targeted model based on an understanding of the underlying biological mechanism of a disease. In this new framework, scientists seek to identify a protein target associated with a disease and develop a molecule that can modulate that protein target. As a shorthand to describe the biological activity of a given molecule, scientists assign a label referred to as mechanism-of-action or MoA for short.\n",
    "\n",
    "### How do we determine the MoAs of a new drug?\n",
    "\n",
    "One approach is to treat a sample of human cells with the drug and then analyze the cellular responses with algorithms that search for similarity to known patterns in large genomic databases, such as libraries of gene expression or cell viability patterns of drugs with known MoAs.\n",
    "\n",
    "In this competition, you will have access to a unique dataset that combines gene expression and cell viability data. The data is based on a new technology that measures simultaneously (within the same samples) human cellsâ€™ responses to drugs in a pool of 100 different cell types (thus solving the problem of identifying ex-ante, which cell types are better suited for a given drug). In addition, you will have access to MoA annotations for more than 5,000 drugs in this dataset.\n",
    "\n",
    "As is customary, the dataset has been split into testing and training subsets. Hence, your task is to use the training dataset to develop an algorithm that automatically labels each case in the test set as one or more MoA classes. Note that since drugs can have multiple MoA annotations, the task is formally a multi-label classification problem."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Data\n",
    "\n",
    "- train_features.csv - Features for the training set. Features g- signify gene expression data, and c- signify cell viability data. cp_type indicates samples treated with a compound (cp_vehicle) or with a control perturbation (ctrl_vehicle); control perturbations have no MoAs; cp_time and cp_dose indicate treatment duration (24, 48, 72 hours) and dose (high or low).\n",
    "- train_targets_scored.csv - The binary MoA targets that are scored.\n",
    "- train_targets_nonscored.csv - Additional (optional) binary MoA responses for the training data. These are not predicted nor scored.\n",
    "- test_features.csv - Features for the test data. You must predict the probability of each scored MoA for each row in the test data.\n",
    "- sample_submission.csv - A submission file in the correct format."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Import Libraries\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np \n",
    "import pandas as pd \n",
    "import os\n",
    "\n",
    "from sklearn.model_selection import KFold\n",
    "from sklearn.metrics import log_loss\n",
    " \n",
    "import tensorflow as tf\n",
    "import tensorflow_addons as tfa\n",
    "import tensorflow.keras.backend as K\n",
    "import tensorflow.keras.layers as L\n",
    "import tensorflow.keras.models as M"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/users/PAS1475/lzqian8023/BMI8050-2020/2_dnn_example\r\n"
     ]
    }
   ],
   "source": [
    "!pwd"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Prepare Dataset\n",
    "\n",
    "### Copy datasets from course project directory if not exist"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "from distutils.dir_util import copy_tree\n",
    "import os.path\n",
    "from os import path\n",
    "if not path.exists('train_features.csv'):\n",
    "    copy_tree(\"/fs/ess/PAS1791/course_code/dnn_example\", \".\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Load datasets, note we only use the first 1000 lines for the example"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "test_df = pd.read_csv('./test_features.csv')\n",
    "train_df = pd.read_csv('./train_features.csv')\n",
    "train_target_df = pd.read_csv('./train_targets_scored.csv')\n",
    "sub = pd.read_csv('./sample_submission.csv')\n",
    "\n",
    "target_cols = train_target_df.columns[1:]\n",
    "N_TARGETS = len(target_cols)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "test_df = test_df.iloc[0:1000:,]\n",
    "train_df = train_df.iloc[0:1000,:]\n",
    "sub = sub.iloc[0:1000,:]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sig_id</th>\n",
       "      <th>cp_type</th>\n",
       "      <th>cp_time</th>\n",
       "      <th>cp_dose</th>\n",
       "      <th>g-0</th>\n",
       "      <th>g-1</th>\n",
       "      <th>g-2</th>\n",
       "      <th>g-3</th>\n",
       "      <th>g-4</th>\n",
       "      <th>g-5</th>\n",
       "      <th>...</th>\n",
       "      <th>c-90</th>\n",
       "      <th>c-91</th>\n",
       "      <th>c-92</th>\n",
       "      <th>c-93</th>\n",
       "      <th>c-94</th>\n",
       "      <th>c-95</th>\n",
       "      <th>c-96</th>\n",
       "      <th>c-97</th>\n",
       "      <th>c-98</th>\n",
       "      <th>c-99</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>id_000644bb2</td>\n",
       "      <td>trt_cp</td>\n",
       "      <td>24</td>\n",
       "      <td>D1</td>\n",
       "      <td>1.0620</td>\n",
       "      <td>0.5577</td>\n",
       "      <td>-0.2479</td>\n",
       "      <td>-0.6208</td>\n",
       "      <td>-0.1944</td>\n",
       "      <td>-1.0120</td>\n",
       "      <td>...</td>\n",
       "      <td>0.2862</td>\n",
       "      <td>0.2584</td>\n",
       "      <td>0.8076</td>\n",
       "      <td>0.5523</td>\n",
       "      <td>-0.1912</td>\n",
       "      <td>0.6584</td>\n",
       "      <td>-0.3981</td>\n",
       "      <td>0.2139</td>\n",
       "      <td>0.3801</td>\n",
       "      <td>0.4176</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>id_000779bfc</td>\n",
       "      <td>trt_cp</td>\n",
       "      <td>72</td>\n",
       "      <td>D1</td>\n",
       "      <td>0.0743</td>\n",
       "      <td>0.4087</td>\n",
       "      <td>0.2991</td>\n",
       "      <td>0.0604</td>\n",
       "      <td>1.0190</td>\n",
       "      <td>0.5207</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.4265</td>\n",
       "      <td>0.7543</td>\n",
       "      <td>0.4708</td>\n",
       "      <td>0.0230</td>\n",
       "      <td>0.2957</td>\n",
       "      <td>0.4899</td>\n",
       "      <td>0.1522</td>\n",
       "      <td>0.1241</td>\n",
       "      <td>0.6077</td>\n",
       "      <td>0.7371</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>id_000a6266a</td>\n",
       "      <td>trt_cp</td>\n",
       "      <td>48</td>\n",
       "      <td>D1</td>\n",
       "      <td>0.6280</td>\n",
       "      <td>0.5817</td>\n",
       "      <td>1.5540</td>\n",
       "      <td>-0.0764</td>\n",
       "      <td>-0.0323</td>\n",
       "      <td>1.2390</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.7250</td>\n",
       "      <td>-0.6297</td>\n",
       "      <td>0.6103</td>\n",
       "      <td>0.0223</td>\n",
       "      <td>-1.3240</td>\n",
       "      <td>-0.3174</td>\n",
       "      <td>-0.6417</td>\n",
       "      <td>-0.2187</td>\n",
       "      <td>-1.4080</td>\n",
       "      <td>0.6931</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>id_0015fd391</td>\n",
       "      <td>trt_cp</td>\n",
       "      <td>48</td>\n",
       "      <td>D1</td>\n",
       "      <td>-0.5138</td>\n",
       "      <td>-0.2491</td>\n",
       "      <td>-0.2656</td>\n",
       "      <td>0.5288</td>\n",
       "      <td>4.0620</td>\n",
       "      <td>-0.8095</td>\n",
       "      <td>...</td>\n",
       "      <td>-2.0990</td>\n",
       "      <td>-0.6441</td>\n",
       "      <td>-5.6300</td>\n",
       "      <td>-1.3780</td>\n",
       "      <td>-0.8632</td>\n",
       "      <td>-1.2880</td>\n",
       "      <td>-1.6210</td>\n",
       "      <td>-0.8784</td>\n",
       "      <td>-0.3876</td>\n",
       "      <td>-0.8154</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>id_001626bd3</td>\n",
       "      <td>trt_cp</td>\n",
       "      <td>72</td>\n",
       "      <td>D2</td>\n",
       "      <td>-0.3254</td>\n",
       "      <td>-0.4009</td>\n",
       "      <td>0.9700</td>\n",
       "      <td>0.6919</td>\n",
       "      <td>1.4180</td>\n",
       "      <td>-0.8244</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0042</td>\n",
       "      <td>0.0048</td>\n",
       "      <td>0.6670</td>\n",
       "      <td>1.0690</td>\n",
       "      <td>0.5523</td>\n",
       "      <td>-0.3031</td>\n",
       "      <td>0.1094</td>\n",
       "      <td>0.2885</td>\n",
       "      <td>-0.3786</td>\n",
       "      <td>0.7125</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 876 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         sig_id cp_type  cp_time cp_dose     g-0     g-1     g-2     g-3  \\\n",
       "0  id_000644bb2  trt_cp       24      D1  1.0620  0.5577 -0.2479 -0.6208   \n",
       "1  id_000779bfc  trt_cp       72      D1  0.0743  0.4087  0.2991  0.0604   \n",
       "2  id_000a6266a  trt_cp       48      D1  0.6280  0.5817  1.5540 -0.0764   \n",
       "3  id_0015fd391  trt_cp       48      D1 -0.5138 -0.2491 -0.2656  0.5288   \n",
       "4  id_001626bd3  trt_cp       72      D2 -0.3254 -0.4009  0.9700  0.6919   \n",
       "\n",
       "      g-4     g-5  ...    c-90    c-91    c-92    c-93    c-94    c-95  \\\n",
       "0 -0.1944 -1.0120  ...  0.2862  0.2584  0.8076  0.5523 -0.1912  0.6584   \n",
       "1  1.0190  0.5207  ... -0.4265  0.7543  0.4708  0.0230  0.2957  0.4899   \n",
       "2 -0.0323  1.2390  ... -0.7250 -0.6297  0.6103  0.0223 -1.3240 -0.3174   \n",
       "3  4.0620 -0.8095  ... -2.0990 -0.6441 -5.6300 -1.3780 -0.8632 -1.2880   \n",
       "4  1.4180 -0.8244  ...  0.0042  0.0048  0.6670  1.0690  0.5523 -0.3031   \n",
       "\n",
       "     c-96    c-97    c-98    c-99  \n",
       "0 -0.3981  0.2139  0.3801  0.4176  \n",
       "1  0.1522  0.1241  0.6077  0.7371  \n",
       "2 -0.6417 -0.2187 -1.4080  0.6931  \n",
       "3 -1.6210 -0.8784 -0.3876 -0.8154  \n",
       "4  0.1094  0.2885 -0.3786  0.7125  \n",
       "\n",
       "[5 rows x 876 columns]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "train_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sig_id</th>\n",
       "      <th>cp_type</th>\n",
       "      <th>cp_time</th>\n",
       "      <th>cp_dose</th>\n",
       "      <th>g-0</th>\n",
       "      <th>g-1</th>\n",
       "      <th>g-2</th>\n",
       "      <th>g-3</th>\n",
       "      <th>g-4</th>\n",
       "      <th>g-5</th>\n",
       "      <th>...</th>\n",
       "      <th>c-90</th>\n",
       "      <th>c-91</th>\n",
       "      <th>c-92</th>\n",
       "      <th>c-93</th>\n",
       "      <th>c-94</th>\n",
       "      <th>c-95</th>\n",
       "      <th>c-96</th>\n",
       "      <th>c-97</th>\n",
       "      <th>c-98</th>\n",
       "      <th>c-99</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>id_0004d9e33</td>\n",
       "      <td>trt_cp</td>\n",
       "      <td>24</td>\n",
       "      <td>D1</td>\n",
       "      <td>-0.5458</td>\n",
       "      <td>0.1306</td>\n",
       "      <td>-0.5135</td>\n",
       "      <td>0.4408</td>\n",
       "      <td>1.5500</td>\n",
       "      <td>-0.1644</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0981</td>\n",
       "      <td>0.7978</td>\n",
       "      <td>-0.1430</td>\n",
       "      <td>-0.2067</td>\n",
       "      <td>-0.2303</td>\n",
       "      <td>-0.1193</td>\n",
       "      <td>0.0210</td>\n",
       "      <td>-0.0502</td>\n",
       "      <td>0.1510</td>\n",
       "      <td>-0.7750</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>id_001897cda</td>\n",
       "      <td>trt_cp</td>\n",
       "      <td>72</td>\n",
       "      <td>D1</td>\n",
       "      <td>-0.1829</td>\n",
       "      <td>0.2320</td>\n",
       "      <td>1.2080</td>\n",
       "      <td>-0.4522</td>\n",
       "      <td>-0.3652</td>\n",
       "      <td>-0.3319</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.1190</td>\n",
       "      <td>-0.1852</td>\n",
       "      <td>-1.0310</td>\n",
       "      <td>-1.3670</td>\n",
       "      <td>-0.3690</td>\n",
       "      <td>-0.5382</td>\n",
       "      <td>0.0359</td>\n",
       "      <td>-0.4764</td>\n",
       "      <td>-1.3810</td>\n",
       "      <td>-0.7300</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>id_002429b5b</td>\n",
       "      <td>ctl_vehicle</td>\n",
       "      <td>24</td>\n",
       "      <td>D1</td>\n",
       "      <td>0.1852</td>\n",
       "      <td>-0.1404</td>\n",
       "      <td>-0.3911</td>\n",
       "      <td>0.1310</td>\n",
       "      <td>-1.4380</td>\n",
       "      <td>0.2455</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.2261</td>\n",
       "      <td>0.3370</td>\n",
       "      <td>-1.3840</td>\n",
       "      <td>0.8604</td>\n",
       "      <td>-1.9530</td>\n",
       "      <td>-1.0140</td>\n",
       "      <td>0.8662</td>\n",
       "      <td>1.0160</td>\n",
       "      <td>0.4924</td>\n",
       "      <td>-0.1942</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>id_00276f245</td>\n",
       "      <td>trt_cp</td>\n",
       "      <td>24</td>\n",
       "      <td>D2</td>\n",
       "      <td>0.4828</td>\n",
       "      <td>0.1955</td>\n",
       "      <td>0.3825</td>\n",
       "      <td>0.4244</td>\n",
       "      <td>-0.5855</td>\n",
       "      <td>-1.2020</td>\n",
       "      <td>...</td>\n",
       "      <td>0.1260</td>\n",
       "      <td>0.1570</td>\n",
       "      <td>-0.1784</td>\n",
       "      <td>-1.1200</td>\n",
       "      <td>-0.4325</td>\n",
       "      <td>-0.9005</td>\n",
       "      <td>0.8131</td>\n",
       "      <td>-0.1305</td>\n",
       "      <td>0.5645</td>\n",
       "      <td>-0.5809</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>id_0027f1083</td>\n",
       "      <td>trt_cp</td>\n",
       "      <td>48</td>\n",
       "      <td>D1</td>\n",
       "      <td>-0.3979</td>\n",
       "      <td>-1.2680</td>\n",
       "      <td>1.9130</td>\n",
       "      <td>0.2057</td>\n",
       "      <td>-0.5864</td>\n",
       "      <td>-0.0166</td>\n",
       "      <td>...</td>\n",
       "      <td>0.4965</td>\n",
       "      <td>0.7578</td>\n",
       "      <td>-0.1580</td>\n",
       "      <td>1.0510</td>\n",
       "      <td>0.5742</td>\n",
       "      <td>1.0900</td>\n",
       "      <td>-0.2962</td>\n",
       "      <td>-0.5313</td>\n",
       "      <td>0.9931</td>\n",
       "      <td>1.8380</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 876 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         sig_id      cp_type  cp_time cp_dose     g-0     g-1     g-2     g-3  \\\n",
       "0  id_0004d9e33       trt_cp       24      D1 -0.5458  0.1306 -0.5135  0.4408   \n",
       "1  id_001897cda       trt_cp       72      D1 -0.1829  0.2320  1.2080 -0.4522   \n",
       "2  id_002429b5b  ctl_vehicle       24      D1  0.1852 -0.1404 -0.3911  0.1310   \n",
       "3  id_00276f245       trt_cp       24      D2  0.4828  0.1955  0.3825  0.4244   \n",
       "4  id_0027f1083       trt_cp       48      D1 -0.3979 -1.2680  1.9130  0.2057   \n",
       "\n",
       "      g-4     g-5  ...    c-90    c-91    c-92    c-93    c-94    c-95  \\\n",
       "0  1.5500 -0.1644  ...  0.0981  0.7978 -0.1430 -0.2067 -0.2303 -0.1193   \n",
       "1 -0.3652 -0.3319  ... -0.1190 -0.1852 -1.0310 -1.3670 -0.3690 -0.5382   \n",
       "2 -1.4380  0.2455  ... -0.2261  0.3370 -1.3840  0.8604 -1.9530 -1.0140   \n",
       "3 -0.5855 -1.2020  ...  0.1260  0.1570 -0.1784 -1.1200 -0.4325 -0.9005   \n",
       "4 -0.5864 -0.0166  ...  0.4965  0.7578 -0.1580  1.0510  0.5742  1.0900   \n",
       "\n",
       "     c-96    c-97    c-98    c-99  \n",
       "0  0.0210 -0.0502  0.1510 -0.7750  \n",
       "1  0.0359 -0.4764 -1.3810 -0.7300  \n",
       "2  0.8662  1.0160  0.4924 -0.1942  \n",
       "3  0.8131 -0.1305  0.5645 -0.5809  \n",
       "4 -0.2962 -0.5313  0.9931  1.8380  \n",
       "\n",
       "[5 rows x 876 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "test_df.head()\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Basic Setup and Helpers\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "SEED = 1234\n",
    "EPOCHS = 4\n",
    "BATCH_SIZE = 16\n",
    "FOLDS = 3\n",
    "REPEATS = 2\n",
    "LR = 0.05\n",
    "N_TARGETS = len(target_cols)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "def seed_everything(seed):\n",
    "    np.random.seed(seed)\n",
    "    os.environ['PYTHONHASHSEED'] = str(seed)\n",
    "    tf.random.set_seed(seed)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "def multi_log_loss(y_true, y_pred):\n",
    "    losses = []\n",
    "    for col in y_true.columns:\n",
    "        losses.append(log_loss(y_true.loc[:, col], y_pred.loc[:, col]))\n",
    "    return np.mean(losses)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Encode Categoricals to Binary\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "def preprocess_df(data):\n",
    "    data['cp_type'] = (data['cp_type'] == 'trt_cp').astype(int)\n",
    "    data['cp_dose'] = (data['cp_dose'] == 'D2').astype(int)\n",
    "    return data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "x_train = preprocess_df(train_df.drop(columns=\"sig_id\"))\n",
    "x_test =preprocess_df(test_df.drop(columns=\"sig_id\"))\n",
    "y_train = train_target_df.drop(columns=\"sig_id\")\n",
    "N_FEATURES = x_train.shape[1]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Define Model Architecture\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "def create_model():\n",
    "    model = tf.keras.Sequential([\n",
    "    tf.keras.layers.Input(N_FEATURES),\n",
    "    tf.keras.layers.BatchNormalization(),\n",
    "    tf.keras.layers.Dropout(0.2),\n",
    "    tfa.layers.WeightNormalization(tf.keras.layers.Dense(2048, activation=\"relu\")),\n",
    "    tf.keras.layers.BatchNormalization(),\n",
    "    tf.keras.layers.Dropout(0.5),\n",
    "    tfa.layers.WeightNormalization(tf.keras.layers.Dense(2048, activation=\"relu\")),\n",
    "    tf.keras.layers.BatchNormalization(),\n",
    "    #tf.keras.layers.Dropout(0.4),\n",
    "    #tfa.layers.WeightNormalization(tf.keras.layers.Dense(2048, activation=\"relu\")),  \n",
    "    #tf.keras.layers.BatchNormalization(),\n",
    "    tf.keras.layers.Dropout(0.5),\n",
    "    tfa.layers.WeightNormalization(tf.keras.layers.Dense(N_TARGETS, activation=\"sigmoid\"))\n",
    "    ])\n",
    "    model.compile(optimizer=tf.keras.optimizers.Adam(learning_rate = LR), loss='binary_crossentropy', metrics=[\"accuracy\"])\n",
    "    return model"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Main CV and Model Training Function\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "def build_train(resume_models = None, repeat_number = 0, folds = 3, skip_folds = 0):\n",
    "    \n",
    "    models = []\n",
    "    oof_preds = y_train.copy()\n",
    "    \n",
    "\n",
    "    kfold = KFold(folds, shuffle = True)\n",
    "    for fold, (train_ind, val_ind) in enumerate(kfold.split(x_train)):\n",
    "        print('\\n')\n",
    "        print('-'*50)\n",
    "        print(f'Training fold {fold + 1}')\n",
    "        \n",
    "        cb_lr_schedule = tf.keras.callbacks.ReduceLROnPlateau(monitor = 'binary_crossentropy', factor = 0.4, patience = 2, verbose = 1, min_delta = 0.0001, mode = 'auto')\n",
    "        checkpoint_path = f'repeat:{repeat_number}_Fold:{fold}.hdf5'\n",
    "        cb_checkpt = tf.keras.callbacks.ModelCheckpoint(checkpoint_path, monitor = 'val_loss', verbose = 0, save_best_only = True, save_weights_only = True, mode = 'min')\n",
    "\n",
    "        model = create_model()\n",
    "        model.fit(x_train.values[train_ind],\n",
    "              y_train.values[train_ind],\n",
    "              validation_data=(x_train.values[val_ind], y_train.values[val_ind]),\n",
    "              callbacks = [cb_lr_schedule, cb_checkpt],\n",
    "              epochs=EPOCHS, batch_size=BATCH_SIZE, verbose=2\n",
    "             )\n",
    "        model.load_weights(checkpoint_path)\n",
    "        oof_preds.loc[val_ind, :] = model.predict(x_train.values[val_ind])\n",
    "        models.append(model)\n",
    "\n",
    "    return models, oof_preds"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "\n",
      "--------------------------------------------------\n",
      "Training fold 1\n",
      "Epoch 1/4\n",
      "WARNING:tensorflow:Reduce LR on plateau conditioned on metric `binary_crossentropy` which is not available. Available metrics are: loss,accuracy,val_loss,val_accuracy,lr\n",
      "42/42 - 4s - loss: 0.1007 - accuracy: 0.0180 - val_loss: 0.0375 - val_accuracy: 0.0120 - lr: 0.0500\n",
      "Epoch 2/4\n",
      "WARNING:tensorflow:Reduce LR on plateau conditioned on metric `binary_crossentropy` which is not available. Available metrics are: loss,accuracy,val_loss,val_accuracy,lr\n",
      "42/42 - 3s - loss: 0.0214 - accuracy: 0.0330 - val_loss: 0.0289 - val_accuracy: 0.0090 - lr: 0.0500\n",
      "Epoch 3/4\n",
      "WARNING:tensorflow:Reduce LR on plateau conditioned on metric `binary_crossentropy` which is not available. Available metrics are: loss,accuracy,val_loss,val_accuracy,lr\n",
      "42/42 - 3s - loss: 0.0204 - accuracy: 0.0300 - val_loss: 0.0240 - val_accuracy: 0.0449 - lr: 0.0500\n",
      "Epoch 4/4\n",
      "WARNING:tensorflow:Reduce LR on plateau conditioned on metric `binary_crossentropy` which is not available. Available metrics are: loss,accuracy,val_loss,val_accuracy,lr\n",
      "42/42 - 3s - loss: 0.0198 - accuracy: 0.0390 - val_loss: 0.0231 - val_accuracy: 0.0509 - lr: 0.0500\n",
      "\n",
      "\n",
      "--------------------------------------------------\n",
      "Training fold 2\n",
      "Epoch 1/4\n",
      "WARNING:tensorflow:Reduce LR on plateau conditioned on metric `binary_crossentropy` which is not available. Available metrics are: loss,accuracy,val_loss,val_accuracy,lr\n",
      "42/42 - 3s - loss: 0.1118 - accuracy: 0.0090 - val_loss: 0.0379 - val_accuracy: 0.0120 - lr: 0.0500\n",
      "Epoch 2/4\n",
      "WARNING:tensorflow:Reduce LR on plateau conditioned on metric `binary_crossentropy` which is not available. Available metrics are: loss,accuracy,val_loss,val_accuracy,lr\n",
      "42/42 - 3s - loss: 0.0217 - accuracy: 0.0285 - val_loss: 0.0342 - val_accuracy: 0.0330 - lr: 0.0500\n",
      "Epoch 3/4\n",
      "WARNING:tensorflow:Reduce LR on plateau conditioned on metric `binary_crossentropy` which is not available. Available metrics are: loss,accuracy,val_loss,val_accuracy,lr\n",
      "42/42 - 3s - loss: 0.0203 - accuracy: 0.0270 - val_loss: 0.0261 - val_accuracy: 0.0420 - lr: 0.0500\n",
      "Epoch 4/4\n",
      "WARNING:tensorflow:Reduce LR on plateau conditioned on metric `binary_crossentropy` which is not available. Available metrics are: loss,accuracy,val_loss,val_accuracy,lr\n",
      "42/42 - 3s - loss: 0.0199 - accuracy: 0.0360 - val_loss: 0.0242 - val_accuracy: 0.0270 - lr: 0.0500\n",
      "\n",
      "\n",
      "--------------------------------------------------\n",
      "Training fold 3\n",
      "Epoch 1/4\n",
      "WARNING:tensorflow:Reduce LR on plateau conditioned on metric `binary_crossentropy` which is not available. Available metrics are: loss,accuracy,val_loss,val_accuracy,lr\n",
      "42/42 - 3s - loss: 0.1242 - accuracy: 0.0240 - val_loss: 0.0402 - val_accuracy: 0.0150 - lr: 0.0500\n",
      "Epoch 2/4\n",
      "WARNING:tensorflow:Reduce LR on plateau conditioned on metric `binary_crossentropy` which is not available. Available metrics are: loss,accuracy,val_loss,val_accuracy,lr\n",
      "42/42 - 3s - loss: 0.0214 - accuracy: 0.0225 - val_loss: 0.0269 - val_accuracy: 0.0360 - lr: 0.0500\n",
      "Epoch 3/4\n",
      "WARNING:tensorflow:Reduce LR on plateau conditioned on metric `binary_crossentropy` which is not available. Available metrics are: loss,accuracy,val_loss,val_accuracy,lr\n",
      "42/42 - 3s - loss: 0.0203 - accuracy: 0.0360 - val_loss: 0.0226 - val_accuracy: 0.0210 - lr: 0.0500\n",
      "Epoch 4/4\n",
      "WARNING:tensorflow:Reduce LR on plateau conditioned on metric `binary_crossentropy` which is not available. Available metrics are: loss,accuracy,val_loss,val_accuracy,lr\n",
      "42/42 - 3s - loss: 0.0198 - accuracy: 0.0240 - val_loss: 0.0226 - val_accuracy: 0.0150 - lr: 0.0500\n",
      "\n",
      "\n",
      "--------------------------------------------------\n",
      "Training fold 1\n",
      "Epoch 1/4\n",
      "WARNING:tensorflow:Reduce LR on plateau conditioned on metric `binary_crossentropy` which is not available. Available metrics are: loss,accuracy,val_loss,val_accuracy,lr\n",
      "42/42 - 3s - loss: 0.1001 - accuracy: 0.0165 - val_loss: 0.0261 - val_accuracy: 0.0629 - lr: 0.0500\n",
      "Epoch 2/4\n",
      "WARNING:tensorflow:Reduce LR on plateau conditioned on metric `binary_crossentropy` which is not available. Available metrics are: loss,accuracy,val_loss,val_accuracy,lr\n",
      "42/42 - 2s - loss: 0.0217 - accuracy: 0.0240 - val_loss: 0.0302 - val_accuracy: 0.0180 - lr: 0.0500\n",
      "Epoch 3/4\n",
      "WARNING:tensorflow:Reduce LR on plateau conditioned on metric `binary_crossentropy` which is not available. Available metrics are: loss,accuracy,val_loss,val_accuracy,lr\n",
      "42/42 - 3s - loss: 0.0205 - accuracy: 0.0195 - val_loss: 0.0257 - val_accuracy: 0.0329 - lr: 0.0500\n",
      "Epoch 4/4\n",
      "WARNING:tensorflow:Reduce LR on plateau conditioned on metric `binary_crossentropy` which is not available. Available metrics are: loss,accuracy,val_loss,val_accuracy,lr\n",
      "42/42 - 3s - loss: 0.0197 - accuracy: 0.0330 - val_loss: 0.0223 - val_accuracy: 0.0659 - lr: 0.0500\n",
      "\n",
      "\n",
      "--------------------------------------------------\n",
      "Training fold 2\n",
      "Epoch 1/4\n",
      "WARNING:tensorflow:Reduce LR on plateau conditioned on metric `binary_crossentropy` which is not available. Available metrics are: loss,accuracy,val_loss,val_accuracy,lr\n",
      "42/42 - 3s - loss: 0.1045 - accuracy: 0.0240 - val_loss: 0.0325 - val_accuracy: 0.0270 - lr: 0.0500\n",
      "Epoch 2/4\n",
      "WARNING:tensorflow:Reduce LR on plateau conditioned on metric `binary_crossentropy` which is not available. Available metrics are: loss,accuracy,val_loss,val_accuracy,lr\n",
      "42/42 - 3s - loss: 0.0225 - accuracy: 0.0180 - val_loss: 0.0267 - val_accuracy: 0.0180 - lr: 0.0500\n",
      "Epoch 3/4\n",
      "WARNING:tensorflow:Reduce LR on plateau conditioned on metric `binary_crossentropy` which is not available. Available metrics are: loss,accuracy,val_loss,val_accuracy,lr\n",
      "42/42 - 3s - loss: 0.0211 - accuracy: 0.0240 - val_loss: 0.0229 - val_accuracy: 0.0150 - lr: 0.0500\n",
      "Epoch 4/4\n",
      "WARNING:tensorflow:Reduce LR on plateau conditioned on metric `binary_crossentropy` which is not available. Available metrics are: loss,accuracy,val_loss,val_accuracy,lr\n",
      "42/42 - 3s - loss: 0.0205 - accuracy: 0.0375 - val_loss: 0.0216 - val_accuracy: 0.0120 - lr: 0.0500\n",
      "\n",
      "\n",
      "--------------------------------------------------\n",
      "Training fold 3\n",
      "Epoch 1/4\n",
      "WARNING:tensorflow:Reduce LR on plateau conditioned on metric `binary_crossentropy` which is not available. Available metrics are: loss,accuracy,val_loss,val_accuracy,lr\n",
      "42/42 - 3s - loss: 0.0946 - accuracy: 0.0120 - val_loss: 0.0384 - val_accuracy: 0.0270 - lr: 0.0500\n",
      "Epoch 2/4\n",
      "WARNING:tensorflow:Reduce LR on plateau conditioned on metric `binary_crossentropy` which is not available. Available metrics are: loss,accuracy,val_loss,val_accuracy,lr\n",
      "42/42 - 3s - loss: 0.0210 - accuracy: 0.0300 - val_loss: 0.0317 - val_accuracy: 0.0180 - lr: 0.0500\n",
      "Epoch 3/4\n",
      "WARNING:tensorflow:Reduce LR on plateau conditioned on metric `binary_crossentropy` which is not available. Available metrics are: loss,accuracy,val_loss,val_accuracy,lr\n",
      "42/42 - 3s - loss: 0.0192 - accuracy: 0.0375 - val_loss: 0.0296 - val_accuracy: 0.0180 - lr: 0.0500\n",
      "Epoch 4/4\n",
      "WARNING:tensorflow:Reduce LR on plateau conditioned on metric `binary_crossentropy` which is not available. Available metrics are: loss,accuracy,val_loss,val_accuracy,lr\n",
      "42/42 - 3s - loss: 0.0187 - accuracy: 0.0525 - val_loss: 0.0266 - val_accuracy: 0.0270 - lr: 0.0500\n"
     ]
    }
   ],
   "source": [
    "models = []\n",
    "oof_preds = []\n",
    "# seed everything\n",
    "seed_everything(SEED)\n",
    "for i in range(REPEATS):\n",
    "    m, oof = build_train(repeat_number = i, folds=FOLDS)\n",
    "    models = models + m\n",
    "    oof_preds.append(oof)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Repeat 1 OOF Log Loss: 0.0009790193899054049\n",
      "Repeat 2 OOF Log Loss: 0.0009863856255755705\n",
      "Mean OOF Log Loss: 0.0009211774053384918\n"
     ]
    }
   ],
   "source": [
    "mean_oof_preds = y_train.copy()\n",
    "mean_oof_preds.loc[:, target_cols] = 0\n",
    "for i, p in enumerate(oof_preds):\n",
    "    print(f\"Repeat {i + 1} OOF Log Loss: {multi_log_loss(y_train, p)}\")\n",
    "    mean_oof_preds.loc[:, target_cols] += p[target_cols]\n",
    "\n",
    "mean_oof_preds.loc[:, target_cols] /= len(oof_preds)\n",
    "print(f\"Mean OOF Log Loss: {multi_log_loss(y_train, mean_oof_preds)}\")\n",
    "#mean_oof_preds.loc[x_train['cp_type'] == 0, target_cols] = 0\n",
    "#print(f\"Mean OOF Log Loss (ctl adjusted): {multi_log_loss(y_train, mean_oof_preds)}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Make Test Predictions and Save SubmissionÂ¶\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "test_preds = sub.copy()\n",
    "test_preds[target_cols] = 0\n",
    "for model in models:\n",
    "    test_preds.loc[:,target_cols] += model.predict(x_test)\n",
    "test_preds.loc[:,target_cols] /= len(models)\n",
    "test_preds.loc[x_test['cp_type'] == 0, target_cols] = 0\n",
    "test_preds.to_csv('submission.csv', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sig_id</th>\n",
       "      <th>5-alpha_reductase_inhibitor</th>\n",
       "      <th>11-beta-hsd1_inhibitor</th>\n",
       "      <th>acat_inhibitor</th>\n",
       "      <th>acetylcholine_receptor_agonist</th>\n",
       "      <th>acetylcholine_receptor_antagonist</th>\n",
       "      <th>acetylcholinesterase_inhibitor</th>\n",
       "      <th>adenosine_receptor_agonist</th>\n",
       "      <th>adenosine_receptor_antagonist</th>\n",
       "      <th>adenylyl_cyclase_activator</th>\n",
       "      <th>...</th>\n",
       "      <th>tropomyosin_receptor_kinase_inhibitor</th>\n",
       "      <th>trpv_agonist</th>\n",
       "      <th>trpv_antagonist</th>\n",
       "      <th>tubulin_inhibitor</th>\n",
       "      <th>tyrosine_kinase_inhibitor</th>\n",
       "      <th>ubiquitin_specific_protease_inhibitor</th>\n",
       "      <th>vegfr_inhibitor</th>\n",
       "      <th>vitamin_b</th>\n",
       "      <th>vitamin_d_receptor_agonist</th>\n",
       "      <th>wnt_inhibitor</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>id_0004d9e33</td>\n",
       "      <td>0.000052</td>\n",
       "      <td>0.002101</td>\n",
       "      <td>0.000060</td>\n",
       "      <td>0.012856</td>\n",
       "      <td>0.024077</td>\n",
       "      <td>0.007482</td>\n",
       "      <td>0.000028</td>\n",
       "      <td>0.004009</td>\n",
       "      <td>0.000055</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000421</td>\n",
       "      <td>0.000484</td>\n",
       "      <td>0.002996</td>\n",
       "      <td>0.007242</td>\n",
       "      <td>0.003786</td>\n",
       "      <td>0.000021</td>\n",
       "      <td>0.006203</td>\n",
       "      <td>0.004342</td>\n",
       "      <td>0.000779</td>\n",
       "      <td>0.005179</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>id_001897cda</td>\n",
       "      <td>0.000038</td>\n",
       "      <td>0.001911</td>\n",
       "      <td>0.000040</td>\n",
       "      <td>0.006325</td>\n",
       "      <td>0.009486</td>\n",
       "      <td>0.006528</td>\n",
       "      <td>0.000017</td>\n",
       "      <td>0.003571</td>\n",
       "      <td>0.000044</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000083</td>\n",
       "      <td>0.000825</td>\n",
       "      <td>0.001397</td>\n",
       "      <td>0.004790</td>\n",
       "      <td>0.001694</td>\n",
       "      <td>0.000011</td>\n",
       "      <td>0.003683</td>\n",
       "      <td>0.002479</td>\n",
       "      <td>0.000397</td>\n",
       "      <td>0.003628</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>id_002429b5b</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>id_00276f245</td>\n",
       "      <td>0.000035</td>\n",
       "      <td>0.001631</td>\n",
       "      <td>0.000038</td>\n",
       "      <td>0.006782</td>\n",
       "      <td>0.016265</td>\n",
       "      <td>0.005603</td>\n",
       "      <td>0.000018</td>\n",
       "      <td>0.003288</td>\n",
       "      <td>0.000040</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000099</td>\n",
       "      <td>0.000469</td>\n",
       "      <td>0.001263</td>\n",
       "      <td>0.006445</td>\n",
       "      <td>0.002219</td>\n",
       "      <td>0.000013</td>\n",
       "      <td>0.005101</td>\n",
       "      <td>0.002591</td>\n",
       "      <td>0.000460</td>\n",
       "      <td>0.004075</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>id_0027f1083</td>\n",
       "      <td>0.000021</td>\n",
       "      <td>0.000110</td>\n",
       "      <td>0.000015</td>\n",
       "      <td>0.027175</td>\n",
       "      <td>0.018693</td>\n",
       "      <td>0.002096</td>\n",
       "      <td>0.000020</td>\n",
       "      <td>0.007207</td>\n",
       "      <td>0.000006</td>\n",
       "      <td>...</td>\n",
       "      <td>0.001759</td>\n",
       "      <td>0.000630</td>\n",
       "      <td>0.000891</td>\n",
       "      <td>0.001360</td>\n",
       "      <td>0.004947</td>\n",
       "      <td>0.000007</td>\n",
       "      <td>0.000716</td>\n",
       "      <td>0.004937</td>\n",
       "      <td>0.000255</td>\n",
       "      <td>0.003849</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 207 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         sig_id  5-alpha_reductase_inhibitor  11-beta-hsd1_inhibitor  \\\n",
       "0  id_0004d9e33                     0.000052                0.002101   \n",
       "1  id_001897cda                     0.000038                0.001911   \n",
       "2  id_002429b5b                     0.000000                0.000000   \n",
       "3  id_00276f245                     0.000035                0.001631   \n",
       "4  id_0027f1083                     0.000021                0.000110   \n",
       "\n",
       "   acat_inhibitor  acetylcholine_receptor_agonist  \\\n",
       "0        0.000060                        0.012856   \n",
       "1        0.000040                        0.006325   \n",
       "2        0.000000                        0.000000   \n",
       "3        0.000038                        0.006782   \n",
       "4        0.000015                        0.027175   \n",
       "\n",
       "   acetylcholine_receptor_antagonist  acetylcholinesterase_inhibitor  \\\n",
       "0                           0.024077                        0.007482   \n",
       "1                           0.009486                        0.006528   \n",
       "2                           0.000000                        0.000000   \n",
       "3                           0.016265                        0.005603   \n",
       "4                           0.018693                        0.002096   \n",
       "\n",
       "   adenosine_receptor_agonist  adenosine_receptor_antagonist  \\\n",
       "0                    0.000028                       0.004009   \n",
       "1                    0.000017                       0.003571   \n",
       "2                    0.000000                       0.000000   \n",
       "3                    0.000018                       0.003288   \n",
       "4                    0.000020                       0.007207   \n",
       "\n",
       "   adenylyl_cyclase_activator  ...  tropomyosin_receptor_kinase_inhibitor  \\\n",
       "0                    0.000055  ...                               0.000421   \n",
       "1                    0.000044  ...                               0.000083   \n",
       "2                    0.000000  ...                               0.000000   \n",
       "3                    0.000040  ...                               0.000099   \n",
       "4                    0.000006  ...                               0.001759   \n",
       "\n",
       "   trpv_agonist  trpv_antagonist  tubulin_inhibitor  \\\n",
       "0      0.000484         0.002996           0.007242   \n",
       "1      0.000825         0.001397           0.004790   \n",
       "2      0.000000         0.000000           0.000000   \n",
       "3      0.000469         0.001263           0.006445   \n",
       "4      0.000630         0.000891           0.001360   \n",
       "\n",
       "   tyrosine_kinase_inhibitor  ubiquitin_specific_protease_inhibitor  \\\n",
       "0                   0.003786                               0.000021   \n",
       "1                   0.001694                               0.000011   \n",
       "2                   0.000000                               0.000000   \n",
       "3                   0.002219                               0.000013   \n",
       "4                   0.004947                               0.000007   \n",
       "\n",
       "   vegfr_inhibitor  vitamin_b  vitamin_d_receptor_agonist  wnt_inhibitor  \n",
       "0         0.006203   0.004342                    0.000779       0.005179  \n",
       "1         0.003683   0.002479                    0.000397       0.003628  \n",
       "2         0.000000   0.000000                    0.000000       0.000000  \n",
       "3         0.005101   0.002591                    0.000460       0.004075  \n",
       "4         0.000716   0.004937                    0.000255       0.003849  \n",
       "\n",
       "[5 rows x 207 columns]"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "test_preds.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "bmi8050",
   "language": "python",
   "name": "bmi8050"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.11"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
